octue/cloud/deployment/google/deployer.py

```diff
@@ -0,0 +1,262 @@
+        self.project_name = octue_configuration["project_name"]
+        self.region = octue_configuration["region"]
+
+        # Generated attributes.
+        self.service_id = f"{OCTUE_NAMESPACE}.{service_id or uuid.uuid4()}"
+        self.build_trigger_description = f"Build {self.name} service and deploy it to Cloud Run."
+
+        self._default_image_uri = (
+            f"{DOCKER_REGISTRY_URL}/{self.project_name}/{self.repository_owner}/{self.repository_name}/"
+            f"{self.name}:{self._get_short_head_commit_hash()}"
+        )
+
+        # Optional configuration file entries.
+        self.dockerfile_path = octue_configuration.get("dockerfile_path")
+        self.minimum_instances = octue_configuration.get("minimum_instances", 0)
+        self.maximum_instances = octue_configuration.get("maximum_instances", 10)
+        self.concurrency = octue_configuration.get("concurrency", 80)
+        self.image_uri = octue_configuration.get("image_uri", self._default_image_uri)
+        self.branch_pattern = octue_configuration.get("branch_pattern", "^main$")
+        self.memory = octue_configuration.get("memory", "128Mi")
+        self.cpus = octue_configuration.get("cpus", 1)
+        self.environment_variables = octue_configuration.get("environment_variables", [])
+
+    def deploy(self, no_cache=False, update=False):
+        total_number_of_stages = 4
+
+        with ProgressMessage("Generating Google Cloud Build configuration", 1, total_number_of_stages):
+            self._generate_cloud_build_configuration(with_cache=not no_cache)
+
+        with tempfile.NamedTemporaryFile() as temporary_file:
+            with open(temporary_file.name, "w") as f:
+                yaml.dump(self.cloud_build_configuration, f)
+
+            with ProgressMessage("Creating build trigger", 2, total_number_of_stages) as progress_message:
+                try:
+                    self._create_build_trigger(cloud_build_configuration_path=temporary_file.name)
+                except subprocess.SubprocessError as e:
+                    self._raise_or_ignore_already_exists_error(e, update, progress_message)
+
+            with ProgressMessage("Building and deploying service", 3, total_number_of_stages):
+                self._build_and_deploy_service(cloud_build_configuration_path=temporary_file.name)
+
+        with ProgressMessage("Creating Eventarc Pub/Sub run trigger", 4, total_number_of_stages) as progress_message:
+            try:
+                self._create_eventarc_run_trigger()
+            except subprocess.SubprocessError as e:
+                self._raise_or_ignore_already_exists_error(e, update, progress_message)
+
+        print(f"[SUCCESS] Service deployed - it can be questioned via Pub/Sub at {self.service_id!r}.")
+
+        return self.service_id
+
+    @staticmethod
+    def _get_short_head_commit_hash():
+        return subprocess.run(["git", "rev-parse", "--short", "HEAD"], capture_output=True).stdout.decode().strip()
```

Comment:
If you run this on a dirty repo (ie with uncommitted changes) it’ll fail so you should handle that with a graceful error.

Reply:
Why would it fail in that case? I just tried it with unstaged and staged changes and it gave me the same output as in the clean repo

Reply:
They must have changed it then; it definitely used to return nothing. Which means actually it’s just incorrect in the event of dirty repo (because you’re not actually at a commit sha)